### PR TITLE
Return the expiration time of the token, depending on its type, on the endpoint of introspection.

### PR DESCRIPTION
**Pull request by pr0head**
## Related issue
https://github.com/ory/hydra/issues/1296

## Proposed changes
The change allows to get the correct expiration time of the token, depending on its type.

## Checklist
- [x] I have read the [contributing guidelines](CONTRIBUTING.md)
- [x] I confirm that this pull request does not address a security vulnerability. If this pull request addresses a security vulnerability, I confirm that I got green light (please contact [hi@ory.sh](mailto:hi@ory.sh)) from the maintainers to push the changes.
- [x] I signed the [Developer's Certificate of Origin](https://github.com/ory/keto/blob/master/CONTRIBUTING.md#developers-certificate-of-origin)
by signing my commit(s). You can amend your signature to the most recent commit by using `git commit --amend -s`. If you amend the commit, you might need to force push using `git push --force HEAD:<branch>`. Please be very careful when using force push.
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation within the code base (if appropriate)
- [ ] I have documented my changes in the [developer guide](https://github.com/ory/docs) (if appropriate)

### CODE REVIEW

**Comment by pr0head** [comment]
Sorry, new PR created

**Comment by pr0head** [comment]
Couldn't do the tests, I'll try later.

File: oauth2/handler.go [hunk 1]
```diff
@@ -430,9 +430,13 @@ func (h *Handler) IntrospectHandler(w http.ResponseWriter, r *http.Request, _ ht
...
-	exp := resp.GetAccessRequester().GetSession().GetExpiresAt(fosite.AccessToken)
+	exp := resp.GetAccessRequester().GetSession().GetExpiresAt(tt)
 	if exp.IsZero() {
-		exp = resp.GetAccessRequester().GetRequestedAt().Add(h.AccessTokenLifespan)
+		if tt == fosite.RefreshToken {
+			exp = resp.GetAccessRequester().GetRequestedAt().Add(h.RefreshTokenLifespan)
+		} else {
+			exp = resp.GetAccessRequester().GetRequestedAt().Add(h.AccessTokenLifespan)
+		}
 	}
```
Comment (aaslamin) [inline oauth2/handler.go]: What about the lifetime of authorization codes and id_tokens? Perhaps fetching the expiry can be extracted to a non-exported helper somewhere within this package or file.

Comment (aaslamin) [reply oauth2/handler.go]: Also, it is possible for refresh tokens to not expire if the value of the lifespan has been set to -1. In that case, perhaps it is best to omit the expiry field in the response payload.

@aeneasr thoughts?

**Comment by aaslamin** [comment]
Thanks for taking this on @pr0head! 🏆

File: oauth2/handler.go (hunk 1 above)
Comment (pr0head) [reply oauth2/handler.go]: > What about the lifetime of authorization codes and id_tokens? Perhaps fetching the expiry can be extracted to a non-exported helper somewhere within this package or file.

The introspection point serves access or update tokens. Other tokens are outside the specification. 
https://tools.ietf.org/html/rfc7662#section-2.1

Comment (aaslamin) [reply oauth2/handler.go]: Ah right! 💯

**Comment by aeneasr** [comment]
Why is `golang.org/x/...` so f* annoying. Use `go mod tidy` to update... :)

**Comment by pr0head** [comment]
> Why is `golang.org/x/...` so f* annoying. Use `go mod tidy` to update... :)

But I did not make any new dependencies, did not change the go.mod and go.sum. Maybe I do not understand something, could you explain to me?

File: oauth2/handler.go (hunk 1 above)
Comment (aeneasr) [reply oauth2/handler.go]: > Also, it is possible for refresh tokens to not expire if the value of the lifespan has been set to -1. In that case, perhaps it is best to omit the expiry field in the response payload.

Damn, you know this system better by me. I'm almost 💯 that's true. In that case it should be omitted.

> The introspection point serves access or update tokens. Other tokens are outside the specification.
https://tools.ietf.org/html/rfc7662#section-2.1

Probably just a typo, introspection serves **access** and **refresh** tokens. So @aaslamin point is right!

**Comment by aeneasr** [comment]
> For this behavioural change we will want tests to avoid future regressions.
> 
> Some test scenarios:
> 
> * Introspecting a refresh token that have been configured to never expire
> * Introspecting a refresh token that has been configured to expire (_use a different expiry than that of the access token_)
> * Introspecting an access token
> * Introspecting an authorization code?
> * Introspecting an id_token?

Only access and refresh tokens are relevant :) All the other cases are super valid though!

**Comment by aeneasr** [comment]
> > Why is `golang.org/x/...` so f* annoying. Use `go mod tidy` to update... :)
> 
> But I did not make any new dependencies, did not change the go.mod and go.sum. Maybe I do not understand something, could you explain to me?

I would love to but I can't. For some reason `golang.org/x`is special and breaks constantly. If you find out why I'd invite you for a night out or something because it's so f* annoying.

**Comment by aeneasr** [comment]
Hi there! Sorry for the silence here - can you rebase on master? The broken test will go away by magic then. :)

**Comment by aeneasr** [comment]
Nice, thanks! :)

**Pull request merged**